src/Components/Components/src/Rendering/Renderer.cs

```diff
@@ -266,9 +296,23 @@ public void DispatchEvent(int componentId, int eventHandlerId, UIEventArgs event
         /// <param name="workItem">The work item to execute.</param>
         public virtual Task Invoke(Action workItem)
         {
-            // Base renderer has nothing to dispatch to, so execute directly
-            workItem();
-            return Task.CompletedTask;
+            if (!(_dispatcher is SynchronizationContext synchronizationContext))
+            {
+                workItem();
+                return Task.CompletedTask;
+            }
+
+            if (SynchronizationContext.Current == synchronizationContext)
```

Comment:
OK I sort of see what's happening. We need the dispatcher to be a `SynchronizationContext` so that we can meaningfully compare it with `SynchronizationContext.Current`.

But does this imply we shouldn't really have an `IDispatcher` interface? With this logic, it's entirely meaningless to implement `IDispatcher` *except* when also inheriting from `SynchronizationContext`. So would it be better to eliminate `IDispatcher` and just use `SynchronizationContext` as the type for `_dispatcher`? Or maybe some concreate class derived from `SynchronizationContext`?

Reply:
You could potentially have a dispatcher that is not a sync context, with the limitation that all async work should be done through dispatcher.Invoke and dispatcher.InvokeAsync. In real life, I think its most likely that it's also going to be a synchronization context, however I feel bad about offering that interface to users.

Neither WPF nor WinForms do so (they have alternatives like Application.Current.Dispatcher) I think it's just fine if we internally handle the fact that is a synchronization context, part of the reason is done this way is because it's been lifted up from remoterenderer and I didn't want to change the semantics if not needed.

I'll rewrite some things to make the implementation clearer and add some comments